### PR TITLE
Bluetooth: hci_raw: Fix pushing H4 headers in bt_buf_get_rx

### LINKED ISSUE #2955: Use interrupt-driven TX in hci_uart sample
**_Reported by Carles Cufi:_**

The current hci_uart sample uses uart_poll_out() to send bytes out to the Host. It should instead use uart_fifo_fill() and wait on a semaphore until there's space in the UART FIFO to send some more if it's full.

(Imported from Jira ZEP-1485)

### LINKED ISSUE #24646: Bluetooth: hci_uart broken on master
Today we bisected an issue with HCI events coming from the hci_uart controller being corrupted (i.e. spurious data) to c66bb21a35c80190686efad26b9628cf24691605 

it happens on first connection, very easy to reproduce. Just flash hci_uart onto a Nordic board, use QEMU/native posix as Host and then start advertising on the shell, connect from any peer central to it.
It typically happens always after the first connection is established.

This is the sort of thing that you see on btmon:

```
> HCI Event: LE Meta Event (0x3e) plen 31                                                    #3025 [hci0] 3141.057741
      LE Enhanced Connection Complete (0x0a)
        Status: Success (0x00)
        Handle: 0
        Role: Slave (0x01)
        Peer address type: Public (0x00)
        Peer address: 00:28:F8:B3:4D:08 (Intel Corporate)
        Local resolvable private address: 7E:6C:28:71:4C:C3 (Resolvable)
        Peer resolvable private address: 00:00:00:00:00:00 (Non-Resolvable)
        Connection interval: 56.25 msec (0x002d)
        Connection latency: 0 (0x0000)
        Supervision timeout: 9600 msec (0x03c0)
        Master clock accuracy: 0x01
< HCI Command: LE Read Remote Used Features (0x08|0x0016) plen 2                             #3026 [hci0] 3141.065027
        Handle: 0
> HCI Event: Connect Request (0x04) plen 15                                                  #3027 [hci0] 3141.139494
        invalid packet size
        04 00 01 16 20 04 04 3e 0c 04 00 00 00 ff 00     .... ..>.......
```

### PR DESCRIPTION
**Pull request by Vudentz**
User of bt_buf_get_rx may attempt to push their own headers, e.g. ACL
headers, so move the H4 header logic to bt_recv after logging into the
monitor since otherwise the buffer would contain H4 headers which is
not expected by the monitor.
    
Fixes #24646

### CODE REVIEW

File: samples/bluetooth/hci_uart/src/main.c [hunk 1]
```diff
@@ -110,6 +112,30 @@ static void h4_acl_recv(struct net_buf *buf, int *remaining)
...
+
+	bytes = uart_fifo_fill(hci_uart_dev, buf->data, buf->len);
+	net_buf_pull(buf, bytes);
+
+	if (buf->len) {
+		return;
+	}
+
+	net_buf_unref(buf);
+	buf = NULL;
```
Comment (jhedberg) [inline samples/bluetooth/hci_uart/src/main.c:136]: I'd just make this:
```
if (!buf->len) {
        net_buf_unref(buf);
        buf = NULL;
}

**Review by jhedberg — CHANGES_REQUESTED** [review]
One minor stylistic comment, otherwise looks good. Hopefully it's tested?

**Review by carlescufi — CHANGES_REQUESTED** [review]
So, of the 2 commits:

`Bluetooth: hci_raw: Fix pushing H4 headers in bt_buf_get_rx` fixes the issue

`Bluetooth: hci_uart: Use interrupt to trigger TX data` makes `hci_uart` not even respond to the first command.

Seeing how we are close to release, I would suggest dropping `Bluetooth: hci_uart: Use interrupt to trigger TX data`  from this PR. 

@jhedberg @Vudentz @joerchan thoughts?

**Comment by Vudentz** [comment]
> So, of the 2 commits:
> 
> `Bluetooth: hci_raw: Fix pushing H4 headers in bt_buf_get_rx` fixes the issue
> 
> `Bluetooth: hci_uart: Use interrupt to trigger TX data` makes `hci_uart` not even respond to the first command.
> 
> Seeing how we are close to release, I would suggest dropping `Bluetooth: hci_uart: Use interrupt to trigger TX data` from this PR.
> 
> @jhedberg @Vudentz @joerchan thoughts?

Sure, the first one was just a tentive to fix the later.

**Review by carlescufi — APPROVED** [review]
Re-tested, works fine, thank you!

**Review by jhedberg — APPROVED** [review]

**Pull request merged**